decrypt.go

```diff
@@ -41,6 +66,42 @@ func (p7 *PKCS7) Decrypt(cert *x509.Certificate, pkey crypto.PrivateKey) ([]byte
 	return nil, ErrUnsupportedAlgorithm
 }
 
+// RFC 4055, 4.1
+// The current ASN.1 parser does not support non-integer defaults so the 'default:' tags here do nothing.
+type rsaOAEPAlgorithmParameters struct {
+	HashFunc    pkix.AlgorithmIdentifier `asn1:"optional,explicit,tag:0,default:sha1Identifier"`
+	MaskGenFunc pkix.AlgorithmIdentifier `asn1:"optional,explicit,tag:1,default:mgf1SHA1Identifier"`
+	PSourceFunc pkix.AlgorithmIdentifier `asn1:"optional,explicit,tag:2,default:pSpecifiedEmptyIdentifier"`
+}
+
+func getHashFuncForKeyEncryptionAlgorithm(keyEncryptionAlgorithm pkix.AlgorithmIdentifier) (crypto.Hash, error) {
+	invalidHashFunc := crypto.Hash(0)
+	params := &rsaOAEPAlgorithmParameters{
+		HashFunc: pkix.AlgorithmIdentifier{Algorithm: OIDDigestAlgorithmSHA1}, // set default hash algorithm to SHA1
+	}
+	var rest []byte
+	rest, err := asn1.Unmarshal(keyEncryptionAlgorithm.Parameters.FullBytes, params)
+	if err != nil {
+		return invalidHashFunc, fmt.Errorf("failed unmarshaling key encryption algorithm parameters: %s", err.Error())
+	}
+	if len(rest) != 0 {
+		return invalidHashFunc, errors.New("trailing data after RSAES-OAEP parameters")
+	}
+
+	switch {
+	case params.HashFunc.Algorithm.Equal(OIDDigestAlgorithmSHA1):
+		return crypto.SHA1, nil
+	case params.HashFunc.Algorithm.Equal(OIDDigestAlgorithmSHA256):
+		return crypto.SHA256, nil
+	case params.HashFunc.Algorithm.Equal(OIDDigestAlgorithmSHA384):
+		return crypto.SHA384, nil
+	case params.HashFunc.Algorithm.Equal(OIDDigestAlgorithmSHA512):
+		return crypto.SHA512, nil
```

Comment:
SHA224 seems to be a valid option too, 2.16.840.1.101.3.4.2.4

Reply:
Added in 76fd9de0905ef97ce7dd9b1e9156c0bd550c14d1